bam_ampliconclip.c

```diff
@@ -64,15 +64,170 @@ typedef struct {
+    bed_entry_t *a = (bed_entry_t *) av;
+    bed_entry_t *b = (bed_entry_t *) bv;
     return a->right < b->right ? -1 : (a->right == b->right ? 0 : 1);
 }
 
 
-int load_bed_file_pairs(char *infile, int get_strand, int sort_by_pos,
-                        bed_pair_list_t *pairs, int64_t *longest) {
+int load_bed_file_multi_ref(char *infile, int get_strand, int sort_by_pos, khash_t(bed_list_hash) *bed_lists) {
+    hFILE *fp;
+    int line_count = 0, ret;
+    int64_t left, right;
+    kstring_t line = KS_INITIALIZE;
+    bed_entry_list_t *list;
+    khiter_t bed_itr;
+
+    if ((fp = hopen(infile, "r")) == NULL) {
+        print_error_errno("ampliconclip", "unable to open file %s.", infile);
+        return 1;
+    }
+
+    char ref[256]; // FIXME long enough?
+
+    while (line.l = 0, kgetline(&line, (kgets_func *)hgets, fp) >= 0) {
+        line_count++;
+        int hret;
+        char strand;
+
+        if (line.l == 0 || *line.s == '#') continue;
+        if (strncmp(line.s, "track ", 6) == 0) continue;
+        if (strncmp(line.s, "browser ", 8) == 0) continue;
+
+        if (get_strand) {
+            if (sscanf(line.s, "%255s %"SCNd64" %"SCNd64" %*s %*s %c",
+                       ref, &left, &right, &strand) != 4) {
+                fprintf(stderr, "[ampliconclip] error: bad bed file format in line %d of %s.\n",
+                                    line_count, infile);
+                ret = 1;
+                goto error;
+            }
+        } else {
+            if (sscanf(line.s, "%255s %"SCNd64" %"SCNd64,
+                       ref, &left, &right) != 3) {
+                fprintf(stderr, "[ampliconclip] error: bad bed file format in line %d of %s",
+                                    line_count, infile);
+                ret = 1;
+                goto error;
+            }
+        }
+
+        bed_itr = kh_get(bed_list_hash, bed_lists, ref);
+
+        if (bed_itr == kh_end(bed_lists)) { // new ref entry
+            char *ref_name = strdup(ref); // need a copy for the hash key
+
+            if (!ref_name) {
+                fprintf(stderr, "[amplicon] error: unable to allocate memory for ref name.\n");
+                ret = 1;
+                goto error;
+            }
+
+            bed_itr = kh_put(bed_list_hash, bed_lists, ref_name, &hret); // FIXME ref needs to be a strdup somewhere
+
+            if (hret >= 0) {
+                list = &kh_val(bed_lists, bed_itr);
+
+                // initialise the new hash entry
+                memcpy(list->ref, ref, 256); // FIXME keep internal ref?
+                list->longest = 0;
+                list->size = 256;
+                list->length = 0;
+
+                if ((list->bp = malloc(list->size * sizeof(bed_entry_t))) == NULL) {
+                    fprintf(stderr, "[ampliconclip] error: unable to allocate memory for bed data.\n");
+                    ret = 1;
+                    goto error;
+                }
```

Comment:
For brevity, this could be simplified by setting size 0 and bp to NULL and letting the realloc below do the initial allocation so we merge the error handling and calls.  Ie here:

```
                list->longest = 0;
                list->size = 0;
                list->length = 0;
                list->bp = NULL
```

and below in the realloc stage, something similar to (instead of `list->size *= 2`)

```
    list->size += list->size/2 + 256;
```

Reply:
I can do that.  It really only removes one if statement but it does reduce the number of moving parts.

Reply:
Thanks.  I don't mind hugely either way, but in general I prefer simplicity if possible.  Less to go wrong and easier to understand (mostly).